File: writecommit.c
#include "common.h"
#include "hprintf.h"
#include "writer.h"

#include <assert.h>
#include <ctype.h>
#include <git2/commit.h>
#include <string.h>


/* Escape characters below as HTML 2.0 / XML 1.0, ignore printing '\r', '\n' */
static void xmlencodeline(FILE* fp, const char* s, size_t len) {
	size_t        i = 0;
	unsigned char c;
	unsigned int  codepoint;

	while (i < len && *s) {
		c = (unsigned char) *s;

		/* Handle ASCII characters */
		if (c < 0x80) {
			switch (c) {
				case '<':
					fputs("&lt;", fp);
					break;
				case '>':
					fputs("&gt;", fp);
					break;
				case '\'':
					fputs("&#39;", fp);
					break;
				case '&':
					fputs("&amp;", fp);
					break;
				case '"':
					fputs("&quot;", fp);
					break;
				case 0x0B:
				case 0x0C:
					break;
				case '\r':
				case '\n':
					break; /* ignore LF */
				default:
					if (isprint(c) || isblank(c))
						putc(c, fp);
					else
						fprintf(fp, "&#%d;", c);
			}
			s++;
			i++;
		}
		/* Handle multi-byte UTF-8 sequences */
		else if (c < 0xC0) {
			/* Invalid continuation byte at start, print as is */
			fprintf(fp, "&#%d;", c);
			s++;
			i++;
		} else {
			/* Decode UTF-8 sequence */
			const unsigned char* start     = (unsigned char*) s;
			int                  remaining = 0;

			if (c < 0xE0) {
				/* 2-byte sequence */
				remaining = 1;
				codepoint = c & 0x1F;
			} else if (c < 0xF0) {
				/* 3-byte sequence */
				remaining = 2;
				codepoint = c & 0x0F;
			} else if (c < 0xF8) {
				/* 4-byte sequence */
				remaining = 3;
				codepoint = c & 0x07;
			} else {
				/* Invalid start byte, print as is */
				fprintf(fp, "&#%d;", c);
				s++;
				i++;
				continue;
			}

			/* Process continuation bytes */
			while (remaining-- && *(++s)) {
				c = (unsigned char) *s;
				if ((c & 0xC0) != 0x80) {
					/* Invalid continuation byte, print original bytes as is */
					while (start <= (unsigned char*) s) {
						fprintf(fp, "&#%d;", *start++);
					}
					s++;
					i++;
					break;
				}
				codepoint = (codepoint << 6) | (c & 0x3F);
			}

			if (remaining < 0) {
				/* Successfully decoded UTF-8 character, output as numeric reference */
				fprintf(fp, "&#%u;", codepoint);
				s++;
				i++;
			}
		}
	}
}

static int hasheadfile(const struct repoinfo* info, const char* filename) {
	for (int i = 0; i < info->headfileslen; i++) {
		if (strcmp(info->headfiles[i], filename) == 0) {
			return 1;
		}
	}
	return 0;
}

void writecommit(FILE* fp, const struct repoinfo* info, git_commit* commit,
                 const struct commitinfo* ci, int parentlink, const char* refname) {
	const git_diff_delta* delta;
	const git_diff_hunk*  hunk;
	const git_diff_line*  line;
	git_patch*            patch;
	size_t                nhunks, nhunklines, changed, add, del, total, i, j, k;
	char                  linestr[80];
	int                   c;
	char                  oid[GIT_OID_SHA1_HEXSIZE + 1], parentoid[GIT_OID_SHA1_HEXSIZE + 1];
	const git_signature*  author  = git_commit_author(commit);
	const char*           msg     = git_commit_message(commit);
	const char*           summary = git_commit_summary(commit);

	git_oid_tostr(oid, sizeof(oid), git_commit_id(commit));
	git_oid_tostr(parentoid, sizeof(parentoid), git_commit_parent_id(commit, 0));

	writeheader(fp, info, 1, info->name, "%y in %s", summary, refname);
	fputs("<pre>", fp);

	hprintf(fp, "<b>commit</b> <a href=\"%s.html\">%s</a>\n", oid, oid);

	if (*parentoid) {
		if (parentlink)
			hprintf(fp, "<b>parent</b> <a href=\"%s.html\">%s</a>\n", parentoid, parentoid);
		else
			hprintf(fp, "<b>parent</b> %s\n", parentoid);
	}
	if (author)
		hprintf(fp, "<b>Author:</b> %y &lt;<a href=\"mailto:%y\">%y</a>&gt;\n<b>Date:</b>   %T\n",
		        author->name, author->email, author->email, &author->when);

	if (msg)
		hprintf(fp, "\n%y\n", msg);

	if (!ci->deltas)
		return;

	if (ci->ndeltas > 1000 || ci->addcount > 100000 || ci->delcount > 100000) {
		fputs("Diff is too large, output suppressed.\n", fp);
		return;
	}

	/* diff stat */
	fputs("<b>Diffstat:</b>\n<table>", fp);
	for (i = 0; i < ci->ndeltas; i++) {
		assert(ci->deltas[i].patch);
		delta = git_patch_get_delta(ci->deltas[i].patch);

		switch (delta->status) {
			case GIT_DELTA_ADDED:
				c = 'A';
				break;
			case GIT_DELTA_COPIED:
				c = 'C';
				break;
			case GIT_DELTA_DELETED:
				c = 'D';
				break;
			case GIT_DELTA_MODIFIED:
				c = 'M';
				break;
			case GIT_DELTA_RENAMED:
				c = 'R';
				break;
			case GIT_DELTA_TYPECHANGE:
				c = 'T';
				break;
			default:
				c = ' ';
				break;
		}
		if (c == ' ')
			fprintf(fp, "<tr><td>%c", c);
		else
			fprintf(fp, "<tr><td class=\"%c\">%c", c, c);

		hprintf(fp, "</td><td><a href=\"#h%zu\">%y", i, delta->old_file.path);
		if (strcmp(delta->old_file.path, delta->new_file.path))
			hprintf(fp, " -&gt; \n%y", delta->new_file.path);

		add     = ci->deltas[i].addcount;
		del     = ci->deltas[i].delcount;
		changed = add + del;
		total   = sizeof(linestr) - 2;
		if (changed > total) {
			if (add)
				add = ((double) total / changed * add) + 1;
			if (del)
				del = ((double) total / changed * del) + 1;
		}
		memset(&linestr, '+', add);
		memset(&linestr[add], '-', del);

		fprintf(fp, "</a></td><td class=\"num\">%zu</td><td class=\"expand\"><span class=\"i\">",
		        ci->deltas[i].addcount + ci->deltas[i].delcount);
		fwrite(&linestr, 1, add, fp);
		fputs("</span><span class=\"d\">", fp);
		fwrite(&linestr[add], 1, del, fp);
		fputs("</span></td></tr>\n", fp);
	}
	fprintf(fp, "</table></pre><pre>%zu file%s changed, %zu insertion%s(+), %zu deletion%s(-)\n",
	        ci->ndeltas, ci->ndeltas == 1 ? "" : "s", ci->addcount, ci->addcount == 1 ? "" : "s",
	        ci->delcount, ci->delcount == 1 ? "" : "s");

	fputs("<hr/>", fp);

	for (i = 0; i < ci->ndeltas; i++) {
		patch = ci->deltas[i].patch;
		delta = git_patch_get_delta(patch);

		if (hasheadfile(info, delta->old_file.path))
			hprintf(fp, "<b>diff --git a/<a id=\"h%zu\" href=\"../file/%h.html\">%y</a> ", i,
			        delta->old_file.path, delta->old_file.path);
		else
			hprintf(fp, "<b>diff --git a/%y ", delta->old_file.path);

		if (hasheadfile(info, delta->new_file.path))
			hprintf(fp, "b/<a href=\"../file/%h.html\">%y</a></b>\n", delta->new_file.path,
			        delta->new_file.path);
		else
			hprintf(fp, "b/%y</b>\n", delta->new_file.path);

		if (delta->flags & GIT_DIFF_FLAG_BINARY) {
			fputs("Binary files differ.\n", fp);
			continue;
		}

		nhunks = git_patch_num_hunks(patch);
		for (j = 0; j < nhunks; j++) {
			if (git_patch_get_hunk(&hunk, &nhunklines, patch, j))
				break;

			hprintf(fp, "<a href=\"#h%zu-%zu\" id=\"h%zu-%zu\" class=\"h\">%y</a>\n", i, j, i, j,
			        hunk->header);

			for (k = 0;; k++) {
				if (git_patch_get_line_in_hunk(&line, patch, j, k))
					break;
				if (line->old_lineno == -1)
					fprintf(fp, "<a href=\"#h%zu-%zu-%zu\" id=\"h%zu-%zu-%zu\" class=\"i\">+", i, j,
					        k, i, j, k);
				else if (line->new_lineno == -1)
					fprintf(fp, "<a href=\"#h%zu-%zu-%zu\" id=\"h%zu-%zu-%zu\" class=\"d\">-", i, j,
					        k, i, j, k);
				else
					putc(' ', fp);
				xmlencodeline(fp, line->content, line->content_len);
				putc('\n', fp);
				if (line->old_lineno == -1 || line->new_lineno == -1)
					fputs("</a>", fp);
			}
		}
	}

	fputs("</pre>\n", fp);
	writefooter(fp);
}

File: writecommit.go
package gitin

import (
	"fmt"
	"html"
	"io"
	"slices"
	"strings"

	git "github.com/jeffwelling/git2go/v37"
)

// #include "fmt.Fprintf.h"
// #include "writer.h"

// #include <assert.h>
// #include <ctype.h>
// #include <git2/commit.h>
// #include <string.h>

const MAXLINESTR = 80

func hasheadfile(info *repoinfo, filename string) bool { // static int hasheadfile(const struct repoinfo* info, const char* filename) {
	return slices.Contains(info.headfiles, filename)
}

func writecommit(fp io.Writer, info *repoinfo, commit *git.Commit, ci *commitinfo, parentlink bool) error { // void writecommit(io.Writer fp, const struct repoinfo* info, git.Commit* commit,
	// // const struct commitinfo* ci, int parentlink) {
	// var delta *git.Diff_delta // const git.Diff_delta* delta;
	// var hunk *git.Diff_hunk   // const git.Diff_hunk*  hunk;
	// var line *git.Diff_line   // const git.Diff_line*  line;
	// var nhunks uint64
	// var nhunklines uint64
	// var i uint64
	// var j uint64
	// var k uint64         // size_t                nhunks, nhunklines, i, j, k;
	// var linestr [80]byte // char                  linestr[80];
	// var c int            // int                   c;

	author := commit.Author()
	summary := commit.Summary()
	msg := commit.Message()
	oid := commit.Id().String()
	var parentoid string
	if parentcommit := commit.ParentId(0); parentcommit != nil {
		parentoid = parentcommit.String()
	}

	writeheader(fp, info, 1, false, info.name, fmt.Sprintf("%s (%s)", html.EscapeString(summary), oid)) // writeheader(fp, info, 1, 0, info->name, "%y (%s)", summary, oid);
	fmt.Fprintf(fp, "<pre>")                                                                            // fmt.Fprintf("<pre>", fp);

	fmt.Fprintf(fp, "<b>commit</b> <a href=\"%s.html\">%s</a>\n", oid, oid) // fmt.Fprintf(fp, "<b>commit</b> <a href=\"%s.html\">%s</a>\n", oid, oid);

	if parentoid != "" { // if (*parentoid) {
		if parentlink { // if (parentlink)
			fmt.Fprintf(fp, "<b>parent</b> <a href=\"%s.html\">%s</a>\n", parentoid, parentoid) // fmt.Fprintf(fp, "<b>parent</b> <a href=\"%s.html\">%s</a>\n", parentoid, parentoid);
		} else {
			fmt.Fprintf(fp, "<b>parent</b> %s\n", parentoid) // fmt.Fprintf(fp, "<b>parent</b> %s\n", parentoid);
		}
	}
	if author != nil { // if (author)
		fmt.Fprintf(fp, "<b>Author:</b> %s &lt;<a href=\"mailto:%s\">%s</a>&gt;\n<b>Date:</b>   %T\n",
			html.EscapeString(author.Name), html.EscapeString(author.Email), html.EscapeString(author.Email), rfc3339(author.When.UTC())) // author->name, author->email, author->email, author->when);
	}
	if msg != "" { // if (msg)
		fmt.Fprintf(fp, "\n%s\n", html.EscapeString(msg)) // fmt.Fprintf(fp, "\n%y\n", msg);
	}
	if len(ci.Deltas) == 0 { // if (!ci->deltas)
		return nil // return;
	}
	if len(ci.Deltas) > 1000 || ci.Addcount > 100000 || ci.Delcount > 100000 { // if (ci->ndeltas > 1000 || ci->addcount > 100000 || ci->delcount > 100000) {
		fmt.Fprintf(fp, "Diff is too large, output suppressed.\n") // fmt.Fprintf("Diff is too large, output suppressed.\n", fp);
		return nil                                                 // return;
	}

	/* diff stat */
	fmt.Fprintf(fp, "<b>Diffstat:</b>\n<table>") // fmt.Fprintf("<b>Diffstat:</b>\n<table>", fp);
	for i, del := range ci.Deltas {              // for (i = 0; i < ci->ndeltas; i++) {
		delta := del.delta // delta = git.Patch_get_delta(ci->deltas[i].patch);

		var c rune
		switch delta.Status { // switch (delta->status) {
		case git.DeltaAdded:
			c = 'A' // c = 'A';
		case git.DeltaCopied:
			c = 'C' // c = 'C';
		case git.DeltaDeleted:
			c = 'D' // c = 'D';
		case git.DeltaModified:
			c = 'M' // c = 'M';
		case git.DeltaRenamed:
			c = 'R' // c = 'R';
		case git.DeltaTypeChange:
			c = 'T' // c = 'T';
		}
		if c == 0 { // if (c == ' ')
			fmt.Fprintf(fp, "<tr><td> ") // fmt.Fprintf(fp, "<tr><td>%c", c);
		} else {
			fmt.Fprintf(fp, "<tr><td class=\"%c\">%c", c, c) // fmt.Fprintf(fp, "<tr><td class=\"%c\">%c", c, c);
		}
		fmt.Fprintf(fp, "</td><td><a href=\"#h%d\">%s", i, html.EscapeString(delta.OldFile.Path)) // fmt.Fprintf(fp, "</td><td><a href=\"#h%zu\">%y", i, delta->old_file.path);
		if delta.OldFile.Path != delta.NewFile.Path {                                             // if (strcmp(delta->old_file.path, delta->new_file.path))
			fmt.Fprintf(fp, " -&gt; \n%s", html.EscapeString(delta.NewFile.Path)) // fmt.Fprintf(fp, " -&gt; \n%y", delta->new_file.path);
		}
		add := del.Addcount      // add     = ci->deltas[i].addcount;
		deletion := del.Delcount // del     = ci->deltas[i].delcount;

		// var total = unsafe.Sizeof(linestr{}) - 2 // total   = sizeof(linestr) - 2;
		changed := add + deletion // changed = add + del;
		if changed > MAXLINESTR { // if (changed > total) {
			if add > 0 { // if (add)
				add = int(float64(MAXLINESTR*add)/float64(changed)) + 1 // add = ((double) total / changed * add) + 1;
			}
			if deletion > 0 { // if (del)
				deletion = int(float64(MAXLINESTR*deletion)/float64(changed)) + 1 // del = ((double) total / changed * del) + 1;
			}
		}

		fmt.Fprintf(fp, "</a></td><td class=\"num\">%d</td><td class=\"expand\"><span class=\"i\">%s</span><span class=\"d\">%s</span></td></tr>\n",
			del.Addcount+del.Delcount, strings.Repeat("+", add), strings.Repeat("-", deletion)) // ci->deltas[i].addcount + ci->deltas[i].delcount);
	}
	var plFile, plAdd, plDel string
	if len(ci.Deltas) > 0 {
		plFile = "s"
	}
	if ci.Addcount > 0 {
		plAdd = "s"
	}
	if ci.Delcount > 0 {
		plDel = "s"
	}
	fmt.Fprintf(fp, "</table></pre><pre>%d file%s changed, %d insertion%s(+), %d deletion%s(-)\n",
		len(ci.Deltas), plFile, ci.Addcount, plAdd, ci.Delcount, plDel) // ci->delcount, ci->delcount == 1 ? "" : "s");

	fmt.Fprintf(fp, "<hr/>") // fmt.Fprintf("<hr/>", fp);

	for i, del := range ci.Deltas { // for (i = 0; i < ci->ndeltas; i++) {
		delta := del.delta // delta            = git.Patch_get_delta(patch);

		if hasheadfile(info, delta.OldFile.Path) { // if (hasheadfile(info, delta->old_file.path))
			fmt.Fprintf(fp, "<b>diff --git a/<a id=\"h%d\" href=\"../file/%s.html\">%s</a> ", i,
				pathunhide(delta.OldFile.Path), html.EscapeString(delta.OldFile.Path)) // delta->old_file.path, delta->old_file.path);
		} else {
			fmt.Fprintf(fp, "<b>diff --git a/%s ", html.EscapeString(delta.OldFile.Path)) // fmt.Fprintf(fp, "<b>diff --git a/%y ", delta->old_file.path);
		}
		if hasheadfile(info, delta.NewFile.Path) { // if (hasheadfile(info, delta->new_file.path))
			fmt.Fprintf(fp, "b/<a href=\"../file/%s.html\">%s</a></b>\n", pathunhide(delta.NewFile.Path), html.EscapeString(delta.NewFile.Path))
		} else {
			fmt.Fprintf(fp, "b/%s</b>\n", html.EscapeString(delta.NewFile.Path)) // fmt.Fprintf(fp, "b/%y</b>\n", delta->new_file.path);
		}
		if delta.Flags&git.DiffFlagBinary > 0 { // if (delta->flags & git.DIFF_FLAG_BINARY) {
			fmt.Fprintf(fp, "Binary files differ.\n") // fmt.Fprintf("Binary files differ.\n", fp);
			continue                                  // continue;
		}

		err := ci.diff.ForEach(func(file git.DiffDelta, progress float64) (git.DiffForEachHunkCallback, error) {
			j := 0
			return func(hunk git.DiffHunk) (git.DiffForEachLineCallback, error) {
				fmt.Fprintf(fp, "<a href=\"#h%d-%d\" id=\"h%d-%d\" class=\"h\">%s</a>\n", i, j, i, j, html.EscapeString(hunk.Header)) // hunk->header);

				j++
				k := 0
				return func(line git.DiffLine) error {
					if line.OldLineno == -1 { // if (line->old_lineno == -1)
						fmt.Fprintf(fp, "<a href=\"#h%d-%d-%d\" id=\"h%d-%d-%d\" class=\"i\">+", i, j, k, i, j, k) // k, i, j, k);
					} else if line.NewLineno == -1 { // else if (line->new_lineno == -1)
						fmt.Fprintf(fp, "<a href=\"#h%d-%d-%d\" id=\"h%d-%d-%d\" class=\"d\">-", i, j, k, i, j, k) // k, i, j, k);
					} else {
						fmt.Fprintf(fp, " ")
					}
					fmt.Fprintf(fp, "%s\n", html.EscapeString(line.Content))
					if line.OldLineno == -1 || line.NewLineno == -1 { // if (line->old_lineno == -1 || line->new_lineno == -1)
						fmt.Fprintf(fp, "</a>") // fmt.Fprintf("</a>", fp);
					}

					k++
					return nil
				}, nil
			}, nil
		}, git.DiffDetailLines)
		if err != nil {
			return err
		}
	}

	fmt.Fprintf(fp, "</pre>\n") // fmt.Fprintf("</pre>\n", fp);
	writefooter(fp)             // writefooter(fp);
	return nil
}
